File: 3553.minimum-weighted-subgraph-with-the-required-paths-ii.py
#
# @lc app=leetcode id=3553 lang=python3
#
# [3553] Minimum Weighted Subgraph With the Required Paths II
#

# @lc code=start
class Solution:
    def minimumWeight(self, edges: List[List[int]], queries: List[List[int]]) -> List[int]:
        from collections import defaultdict
        import heapq
        
        def dijkstra(start):
            dist = {i: float('inf') for i in range(n)}
            dist[start] = 0
            heap = [(0, start)]
            while heap:
                curr_dist, u = heapq.heappop(heap)
                if curr_dist > dist[u]:
                    continue
                for v, weight in graph[u]:
                    if dist[v] > curr_dist + weight:
                        dist[v] = curr_dist + weight
                        heapq.heappush(heap, (dist[v], v))
            return dist
        
        n = len(edges) + 1
        graph = defaultdict(list)
        for u, v, w in edges:
            graph[u].append((v, w))
            graph[v].append((u, w)) # since it's an undirected tree
        
        results = []
        for src1, src2, dest in queries:
            dist_from_src1 = dijkstra(src1)
            dist_from_src2 = dijkstra(src2)
            min_weight = float('inf')
            for node in range(n):
                if node in [src1, src2]:
                    continue
                if dist_from_src1[node] != float('inf') and dist_from_src2[node] != float('inf'):
                    min_weight = min(min_weight, dist_from_src1[node] + dist_from_src2[node])
            result_weight = min_weight + (dist_from_src1[dest] if dest in dist_from_src1 else float('inf')) + (dist_from_src2[dest] if dest in dist_from_src2 else float('inf')) - min_weight # corrects double counting dest path weight once
            results.append(result_weight if result_weight != float('inf') else -1)
        return results
# @lc code=end

File: 3553.minimum-weighted-subgraph-with-the-required-paths-ii.cpp
# \@lc app=leetcode id=3553 lang=cpp
#
# [3553] Minimum Weighted Subgraph With the Required Paths II
#
# \@lc code=start
#include <vector>
#include <queue>
#include <unordered_map>
#include <limits>
using namespace std;

class Solution {
public:
    vector<int> minimumWeight(vector<vector<int>>& edges, vector<vector<int>>& queries) {
        int n = edges.size() + 1; // n is number of nodes since edges are n-1
        vector<unordered_map<int, int>> graph(n);

        // Build graph from edges
        for (const auto& edge : edges) {
            int u = edge[0], v = edge[1], w = edge[2];
            graph[u][v] = w;
            graph[v][u] = w;
        }

        vector<int> result;
        // Lambda function to calculate shortest path using Dijkstra's algorithm from a start node
        auto dijkstra_shortest_path = [&](int start) {
            vector<int> dist(n, numeric_limits<int>::max());
            priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
            pq.push({0, start});
            dist[start] = 0;
            while (!pq.empty()) {
                auto [currentDist, node] = pq.top(); pq.pop();
                if (currentDist > dist[node]) continue;
                for (const auto& [neighbor, weight] : graph[node]) {
                    if (dist[node] + weight < dist[neighbor]) {
                        dist[neighbor] = dist[node] + weight;
                        pq.push({dist[neighbor], neighbor});
                    }
                }
            }
            return dist;
        };

        for (const auto& query : queries) {
data}
data}
data// \@lc code=end